ws-server/src/server.rs

```diff
@@ -91,64 +95,114 @@ impl Server {
 		// Acquire read access to the lock such that additional reader(s) may share this lock.
 		// Write access to this lock will only be possible after the server and all background tasks have stopped.
 		let _stop_handle = self.stop_handle.read().await;
+		let shutdown = self.stop_pair.0;
 
-		let mut incoming = TcpListenerStream::new(self.listener).fuse();
 		let methods = self.methods;
-		let conn_counter = Arc::new(());
 		let mut id = 0;
-		let mut stop_receiver = self.stop_pair.1;
-		let shutdown = self.stop_pair.0;
+
+		let mut driver = ConnDriver::new(self.listener, self.stop_pair.1);
 
 		loop {
-			futures_util::select! {
-				socket = incoming.next() => {
-					if let Some(Ok(socket)) = socket {
-						if let Err(e) = socket.set_nodelay(true) {
-							log::error!("Could not set NODELAY on socket: {:?}", e);
-							continue;
-						}
-
-						if Arc::strong_count(&conn_counter) > self.cfg.max_connections as usize {
-							log::warn!("Too many connections. Try again in a while");
-							continue;
-						}
-
-						let conn_counter2 = conn_counter.clone();
-						let shutdown2 = shutdown.clone();
-						let methods = methods.clone();
-						let cfg = self.cfg.clone();
-						let stop_handle2 = self.stop_handle.clone();
-
-						tokio::spawn(async move {
-							let _ = background_task(socket, id, methods, cfg, shutdown2, stop_handle2).await;
-							drop(conn_counter2);
-						});
-
-						id = id.wrapping_add(1);
-					} else {
-						break;
+			match Pin::new(&mut driver).await {
+				Ok((socket, _addr)) => {
+					if let Err(e) = socket.set_nodelay(true) {
+						log::error!("Could not set NODELAY on socket: {:?}", e);
+						continue;
 					}
-				},
-				stop = stop_receiver.next() => {
-					if stop.is_some() {
-						break;
+
+					if driver.connection_count() >= self.cfg.max_connections as usize {
+						log::warn!("Too many connections. Try again in a while.");
+						continue;
 					}
-				},
-				complete => break,
+
+					let methods = &methods;
+					let cfg = &self.cfg;
+
+					driver.add(Box::pin(handshake(socket, id, methods, cfg, &shutdown, &self.stop_handle)));
+
+					id += 1;
```

Comment:
I would prefer to use `wrapping_add` here but I don't how likely it's that a server reaches 2^32 connections though ^

Reply:
I believe release build will make this into a wrapping add regardless, but might be a good idea to make it explicit.